File: CanTransform777.java
public class CanTransform777 {



    public boolean canTransform(String start, String end) {
        int index = 0, length = start.length();
        while (index < length) {
            char s = start.charAt(index);
            char e = end.charAt(index);
            if (start.charAt(index) == end.charAt(index)) {
                index++;
                continue;
            }
            if (s == 'X') {
                if (e == 'R')
                    return false;
                int ptr = index + 1;
                int offset = -1;
                while (ptr < length) {
                    int c = start.charAt(ptr);
                    if (c == end.charAt(ptr) && c != 'R') {
                        ptr++;
                        continue;
                    }if (start.charAt(ptr) == 'X' && end.charAt(ptr) == 'L')
                        offset--;
                    else if (start.charAt(ptr) == 'L' && end.charAt(ptr) == 'X') {
                        offset++;
                        if (offset > 0)
                            return false;
                    }else
                        break;
                    ptr++;
                }
                if (offset == 0) {
                    index = ptr;
                    continue;
                }
            } else if (s == 'R') {
                if (e == 'L')
                    return false;
                int ptr = index + 1;
                int offset = -1;
                while (ptr < length) {
                    int c = start.charAt(ptr);
                    if (c == end.charAt(ptr) && c != 'L') {
                        ptr++;
                        continue;
                    }
                    if (start.charAt(ptr) == 'R' && end.charAt(ptr) == 'X')
                        offset--;
                    else if (start.charAt(ptr) == 'X' && end.charAt(ptr) == 'R') {
                        offset++;
                        if (offset > 0)
                            return false;
                    }else
                        break;
                    ptr++;
                }
                if (offset == 0) {
                    index = ptr;
                    continue;
                }
            }
            return false;
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println(new CanTransform777().canTransform("XXXXXLXXXLXXXX", "XXLXXXXXXXXLXX"));
    }


}

File: CanTransform777.cpp
#include <string>
#include <iostream>

using namespace std;

class CanTransform777 {
public:
    bool canTransform(string start, string end) {
        int index = 0;
        int offsetL = 0, offsetR = 0;
        while (index < start.length())
        {
            char s = start[index];
            char e = end[index];
            if (s == e)
            {
                if ((s == 'L' && offsetR != 0) || (s == 'R' && offsetL != 0))
                {
                    return false;
                }
                index++;
                continue;
            }
            if ((s == 'L' && e == 'R') || (s == 'R' && e == 'L'))
            {
                return false;
            }
            switch (s)
            {
                case 'X':
                    if (e == 'L')
                    {
                        offsetL--;
                    }else {
                        offsetR++;
                    }
                    break;
                case 'R':
                    offsetR--;
                    break;
                case 'L':
                    offsetL++;
                    break;
                default:
                    break;
            }
            if (offsetR > 0 || offsetL > 0 || (offsetR != 0 && offsetL != 0))
            {
                return false;
            }
            index++;
        }
        if (offsetL == 0 && offsetR == 0)
        {
            return true;
        }
        return false;
    }
};

int main() {
    CanTransform777 transform;
    cout << transform.canTransform("X", "L") << endl;
}